acl-lib/src/cmr/acl/core.clj

```diff
@@ -157,38 +166,81 @@
-  [context permission-type object-identity-type provider-id]
+  target permission in ECHO ACLS for the given permission type."
+  [context permission-type object-identity-type provider-id target]
   ;; Performance optimization here of returning true if it's the system user.
   (or (transmit-config/echo-system-token? context)
       (let [acl-oit-key (access-control/acl-type->acl-key object-identity-type)]
         (->> (get-permitting-acls context
                                   object-identity-type
-                                  "INGEST_MANAGEMENT_ACL"
+                                  target
                                   permission-type)
              ;; Find acls for this provider
              (filter #(or (nil? provider-id)
-                          (= provider-id (get-in % [acl-oit-key :provider-id]))))
+                           (= provider-id (get-in % [acl-oit-key :provider-id]))))
              seq))))
 
+(defn has-subscription-management-permission?
+  "Returns true if the user identified by the token in the cache has been granted
+  EMAIL_SUBSCRIPTION_MANAGEMENT permission in ECHO ACLS for the given permission type."
+  [context permission-type object-identity-type provider-id]
+  (has-management-permission?
+    context permission-type object-identity-type provider-id "EMAIL_SUBSCRIPTION_MANAGEMENT"))
+
+(defn has-ingest-management-permission?
+  "Returns true if the user identified by the token in the cache has been granted
+  INGEST_MANAGEMENT permission in ECHO ACLS for the given permission type."
+  [context permission-type object-identity-type provider-id]
+  (has-management-permission?
+    context permission-type object-identity-type provider-id "INGEST_MANAGEMENT_ACL"))
+
+(defn- verify-management-permission
+  "Verifies the current user has been granted the permission in permission-fn in ECHO ACLs"
+  [context permission-type object-identity-type provider-id cache-key permission-fn]
+  (let [has-permission-fn (fn []
+                            (permission-fn
+                              context permission-type object-identity-type provider-id))
+        has-permission? (if-let [cache (cache/context->cache context cache-key)]
+                          ;; Read using cache. Cache key is combo of token and permission type
+                          (cache/get-value
+                            cache [(:token context) permission-type] has-permission-fn)
+                          ;; No token cache so directly check permission.
+                          (has-permission-fn))]
+    (when-not has-permission?
+      (errors/throw-service-error
+        :unauthorized
+        "You do not have permission to perform that action."))))
+
+(defn verify-subscription-management-permission
+  "Verifies the current user has been granted EMAIL_SUBSCRIPTION_MANAGEMENT
+  permission in ECHO ACLs"
+  ([context]
+   (verify-subscription-management-permission context :update :system-object nil))
```

Comment:
The newly added `EMAIL_SUBSCRIPTION_MANAGEMENT_ACL` is on `provider-object`. It is unnecessary to create these functions on `system-object`.

Reply:
Sorry, cut and paste error. I'll change it to :provider-object